### PR TITLE
Collection expressions: IOperation support

### PR DESCRIPTION
**Pull request by cston**
See API proposal https://github.com/dotnet/roslyn/issues/70631.

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

File: src/Features/CSharpTest/SyncNamespace/SyncNamespaceTests_ChangeNamespace.cs [hunk 1]
```diff
@@ -2343,7 +2343,7 @@ public async Task ChangeNamespace_DoesNotThrowInDuplicateProgramDeclaration()
 
             // No change namespace action because the folder name is not valid identifier
             var (topLevelProgramFolder, topLevelProgramFilePath) = CreateDocumentFilePath(["3B", "C"], "Program.cs");
-            var (duplicateProgramFolder, duplicateProgramFilePath) = CreateDocumentFilePath([], "Program.cs");
+            var (duplicateProgramFolder, _) = CreateDocumentFilePath([], "Program.cs");
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Features/CSharpTest/SyncNamespace/SyncNamespaceTests_ChangeNamespace.cs:2346]: Yes, with this change, an analyzer reports a diagnostic here for an unnecessary assignment.

**Comment by AlekseyTs** [comment]
Are we testing control flow graph with control flow constructs within a collection expression?

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review.

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

**Comment by jcouv** [comment]
Looking

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 2]
```diff
@@ -3387,4 +3387,70 @@
...
+          <para>
+            If the element is an expression, the entry is the expression, with a conversion to
+            the target element type if necessary;
+            otherwise, the entry is an ISpreadOperation.
+          </para>
+        </summary>
+      </Comments>
+    </Property>
+  </Node>
+  <Node Name="ISpreadOperation" Base="IOperation" HasType="true">
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml:3425]: >HasType="true" [](http://example.com/codeflow?start=50&length=14)

That seems unexpected. What's the type of a spread?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 3]
```diff
@@ -6368,6 +6368,46 @@ IArrayInitializerOperation popAndAssembleArrayInitializerValues(IArrayInitialize
...
+            PopStackFrame(frame);
+            return new CollectionExpressionOperation(
+                operation.ConstructMethod,
+                elements,
+                semanticModel: null,
+                operation.Syntax,
+                operation.Type,
+                IsImplicit(operation));
+
+            static IOperation unwrapElement(IOperation element)
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:6384]: >unwrapElement [](http://example.com/codeflow?start=30&length=13)

statis?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 4]
```diff
@@ -6368,6 +6368,46 @@ IArrayInitializerOperation popAndAssembleArrayInitializerValues(IArrayInitialize
             }
         }
 
+        public override IOperation? VisitCollectionExpression(ICollectionExpressionOperation operation, int? argument)
+        {
+            EvalStackFrame frame = PushStackFrame();
+            var elements = VisitArray(operation.Elements, unwrapElement, wrapElement);
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:6374]: >wrapElement [](http://example.com/codeflow?start=73&length=11)

nit: it seems `wrapElement` can't be made static (it uses instance method `IsImplicit`). Consider using a lambda around it to cache the allocated delegate.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 5]
```diff
@@ -6368,6 +6368,46 @@ IArrayInitializerOperation popAndAssembleArrayInitializerValues(IArrayInitialize
             }
         }
 
+        public override IOperation? VisitCollectionExpression(ICollectionExpressionOperation operation, int? argument)
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:6371]: Looking at `VisitInterpolatedStringHandlerCreation` below, I would have expected the CFG for collection expression to surface the nitty-gritty details of the collection evaluation. It sounds like this was discussed/approved in API review, but I'm not clear on whether this is the level of abstraction we want moving forward?

Comment (jcouv) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:6371]: Resolved offline. This is the result of a case-by-case discussion in API review considering the needs of the CFG customer. Thanks

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 6]
```diff
@@ -1221,45 +1219,83 @@ private IArrayInitializerOperation CreateBoundArrayInitializationOperation(Bound
...
+                    BoundCollectionElementInitializer collectionInitializer => collectionInitializer.Arguments[collectionInitializer.InvokedAsExtensionMethod ? 1 : 0],
+                    BoundDynamicCollectionElementInitializer dynamicInitializer => dynamicInitializer.Arguments[0],
                     _ => element,
                 };
-                return Create(result);
             }
         }
 
-        private IOperation CreateBoundCollectionExpressionSpreadElement(BoundCollectionExpressionSpreadElement boundSpreadExpression)
+        private IOperation CreateBoundCollectionExpressionSpreadElement(BoundCollectionExpressionSpreadElement element, BoundExpression? iteratorItem)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1284]: >IOperation [](http://example.com/codeflow?start=16&length=10)

nit: consider declaring a more specific return type

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1284]: Consider doc for `iteratorItem`

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 7]
```diff
@@ -1221,45 +1219,83 @@ private IArrayInitializerOperation CreateBoundArrayInitializationOperation(Bound
...
+        private IOperation CreateBoundCollectionExpressionElement(BoundNode element)
+        {
+            if (element is BoundCollectionExpressionSpreadElement spreadElement)
+            {
+                var iteratorBody = ((BoundExpressionStatement?)spreadElement.IteratorBody)?.Expression;
+                return CreateBoundCollectionExpressionSpreadElement(spreadElement, getUnderlyingExpression(iteratorBody));
+            }
+            return Create(getUnderlyingExpression((BoundExpression)element));
+
+            [return: NotNullIfNotNull("element")] static BoundExpression? getUnderlyingExpression(BoundExpression? element)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1273]: >"element" [](http://example.com/codeflow?start=38&length=9)

nameof(element)?

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml (hunk 2 above)
Comment (cston) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml:3425]: Good catch, thanks.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 6 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1285]: Removed parameter.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:6374]: Changed both local functions to lambda expressions.

**Comment by jcouv** [comment]
>         }

nit: consider adding a test that shows more interesting element conversions than identity or reference conversions. For instance, a collection with different Add methods where each element gets its own conversion.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:21265 in 7320176. [](commit_id = 7320176a2dd1e5bb87ceb26f94f6bdefff90034b, deletion_comment = False)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 9) with some nits to consider

**Comment by cston** [comment]
>         }

Added `CollectionInitializerType_20` with multiple `Add` methods and nullable and boxing conversions.

---
In reply to: [1832759003](https://github.com/dotnet/roslyn/pull/70650#issuecomment-1832759003) [](http://example.com/codeflow?ancestors=1832759003)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:21265 in 7320176. [](commit_id = 7320176a2dd1e5bb87ceb26f94f6bdefff90034b, deletion_comment = False)

**Pull request merged**